### PR TITLE
Concurrent token refresh fix

### PR DESCRIPTION
**Pull request by pavgra**
Fixes https://github.com/OHDSI/WebAPI/issues/436

Solution similar to used in https://github.com/tymondesigns/jwt-auth

### CODE REVIEW

File: src/main/java/org/ohdsi/webapi/shiro/TokenManager.java [hunk 1]
```diff
@@ -1,35 +1,41 @@
...
  */
 public class TokenManager {
-  
+
   private static final String AUTHORIZATION_HEADER = "Authorization";
-  
+
   private static final Map<String, Key> userToKeyMap = new HashMap<>();
-  
+  private static final ExpiringMultimap<String, Key> gracePeriodInvalidTokens = new ExpiringMultimap<>(30000);
```
Comment (chrisknoll) [inline src/main/java/org/ohdsi/webapi/shiro/TokenManager.java]: What does the 30000 ms here in the ExpiringMultimap constructor do?

File: src/main/java/org/ohdsi/webapi/shiro/TokenManager.java [hunk 2]
```diff
@@ -40,32 +46,47 @@ public static String createJsonWebToken(String subject, Date expiration) {
...
+            })
+            .filter(Objects::nonNull)
+            .findFirst()
+            .orElseThrow(() -> new SignatureException("Signing key is not reqistred for the subject."));
   }
-  
+
+  protected static String getUntrustedSubject(String jws) {
+    int i = jws.lastIndexOf('.');
+    String untrustedJwtString = jws.substring(0, i+1);
```
Comment (wivern) [inline src/main/java/org/ohdsi/webapi/shiro/TokenManager.java]: What if string does not contain dot?

File: src/main/java/org/ohdsi/webapi/util/ExpiringMultimap.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        return map.containsKey(key)
+                ? map.get(key).entrySet().stream().map(Map.Entry::getKey).collect(Collectors.toList())
+                : new ArrayList<>();
+    }
+
+
+    public synchronized void put(K key, V value) {
+
+        PassiveExpiringMap<V, Object> storage = map.get(key);
+        if (storage == null) {
```
Comment (wivern) [inline src/main/java/org/ohdsi/webapi/util/ExpiringMultimap.java]: Map.putIfAbsent?

File: src/main/java/org/ohdsi/webapi/shiro/TokenManager.java (hunk 1 above)
Comment (pavgra) [reply src/main/java/org/ohdsi/webapi/shiro/TokenManager.java]: @chrisknoll , defines how long invalidated keys will be stored. I've started moving the value into app's properties, but then payed attention that all methods in the class are static and the class itself is not a @Component, so left it hardcoded for now

**Pull request merged**